cpp/src/iterator/iterator.cuh

```diff
@@ -0,0 +1,322 @@
+  T_output operator()(gdf_index_type i) const {
+    return bit_mask::is_valid(bitmask, i) ? static_cast<T_output>(elements[i]) : identity;
+  }
+};
+
+template <typename T_element, typename T_output>
+struct value_accessor<T_element, T_output, false>
+{
+  T_element const* elements{};
+  value_accessor(T_element const* e, bit_mask::bit_mask_t const*, T_output) : elements{e} {}
+
+  CUDA_HOST_DEVICE_CALLABLE
+  T_output operator()(gdf_index_type i) const { return static_cast<T_output>(elements[i]); }
+};
+
+/** -------------------------------------------------------------------------*
+ * @brief pair accessor with/without null bitmask
+ * A unary function returns pair of scalar value and validity at `id`.
+ * `operator() (gdf_index_type id)` computes
+ * `data` value and validity at `id`
+ * and return `thrust::pair<T_output, bool>(data, validity)`.
+ *
+ * If has_nulls = false, the validity is always true.
+ * If has_nulls = true, the valid flag corresponds to null bitmask flag at `id`,
+ * and the data value = (is_valid(id))? static_cast<T_output>(data[id]) : identity;
+ *
+ * @tparam  T_element cudf data type of input element array
+ * @tparam  T_output  cudf data type of output and `identity` value
+ *                    which is used when null bitmaps flag is false.
+ * @tparam  has_nulls if true, this struct holds only data array.
+ *                    else, this struct holds data array and
+ *                    bitmask array and identity value
+ * -------------------------------------------------------------------------**/
+template <typename T_element, typename T_output, bool has_nulls>
+struct pair_accessor;
+
+template <typename T_element, typename T_output>
+struct pair_accessor<T_element, T_output, true> : public value_accessor<T_element, T_output, true>
+{
+  pair_accessor(T_element const* e, bit_mask::bit_mask_t const* b, T_output i)
+    : value_accessor<T_element, T_output, true>(e, b, i) {};
+
+  CUDA_HOST_DEVICE_CALLABLE
+  thrust::pair<T_output, bool> operator()(gdf_index_type i) const {
+    return bit_mask::is_valid(this->bitmask, i) ?
+        thrust::make_pair(static_cast<T_output>(this->elements[i]), true) :
+        thrust::make_pair(this->identity, false) ;
+  }
+};
+
+template <typename T_element, typename T_output>
+struct pair_accessor<T_element, T_output, false> : public value_accessor<T_element, T_output, false>
+{
+  pair_accessor(T_element const* e, bit_mask::bit_mask_t const* b , T_output i)
+    : value_accessor<T_element, T_output, false>(e, b, i) {};
+
+  CUDA_HOST_DEVICE_CALLABLE
+  thrust::pair<T_output, bool> operator()(gdf_index_type i) const {
+    return thrust::make_pair(static_cast<T_output>(this->elements[i]), true);
+  }
+};
+
+// ---------------------------------------------------------------------------
+// helper functions to make iterator
+
+/** -------------------------------------------------------------------------*
+ * @brief helper function to make a cudf column iterator
```

Comment:
```suggestion
 * @brief Constructs an iterator over the elements of a column.
 * 
 * If the column contains no null values (indicated by `has_nulls == false`) then dereferencing an iterator `it` returned by this function as `*(it + n)` will return `ResultType{ static_cast<ElementType*>(data)[n] }`.
 
 * If the column contains null values (indicated by `has_nulls == true`) then the result of de-referencing an iterator `it` returned by this function as `*(it+n)` will depend if element is valid or null. If the element is valid, it will return `ResultType{ static_cast<ElementType*>(data)[n] }`. If the element is null, it will return `ResultType{identity}`.
```